### PR TITLE
Use an intermediate struct in workload endpoint cache

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

This PR introduces an intermediate struct used in the workload endpoint cache. It will allow us to monitor and synchronize additional wep fields beyond just labels.

This is groundwork for supporting named ports in the policy controller.

### CODE REVIEW

File: pkg/controllers/pod/pod_controller.go [hunk 1]
```diff
@@ -48,27 +48,28 @@ type PodController struct {
...
+		return m, nil
 	}
 
 	cacheArgs := calicocache.ResourceCacheArgs{
 		ListFunc:   listFunc,
-		ObjectType: reflect.TypeOf(map[string]string{}),
+		ObjectType: reflect.TypeOf(converter.WorkloadEndpointData{}),
 	}
 
 	labelCache := calicocache.NewResourceCache(cacheArgs)
```
Comment (bcreane) [inline pkg/controllers/pod/pod_controller.go]: wepCache?

File: pkg/converter/pod_converter.go [hunk 2]
```diff
@@ -26,13 +26,39 @@ import (
...
+	}
+}
+
+// MergeWorkloadEndpointData applies the given WorkloadEndpointData to the provided
+// WorkloadEndpoint, updating relevant fields with new values.
+func MergeWorkloadEndpointData(wep *api.WorkloadEndpoint, upd WorkloadEndpointData) {
+	if wep.Metadata.Workload != upd.Key {
+		log.Fatalf("Bad attempt to merge data for %s into wep %s", upd.Key, wep.Metadata.Workload)
+	}
+	wep.Metadata.Labels = upd.Labels
```
Comment (bcreane) [inline pkg/converter/pod_converter.go]: this is where namedPorts update will happen?

Comment (caseydavenport) [reply pkg/converter/pod_converter.go]: Yep

**Comment by caseydavenport** [comment]
> Seems like we're using a DeepEqual to compare old/new, and then just (currently) overwriting the label field. Shouldn't the old/new compare be limited to the fields we're updating?

Not sure I understand the question - ff there's any difference between the old / new objects, then we need to reprogram the workload endpoint to the datastore in its entirety.

**Comment by bcreane** [comment]
The code looks like its comparing all fields recursively in WorkloadEndpoint, but on line 54 of pod_converter.go, it's only merging/overwriting Labels field.

My question/concern is whether we're comparing more than we're updating. If we compare all fields, but only update one, we could enter a compare/update loop.

**Comment by caseydavenport** [comment]
> If we compare all fields, but only update one, we could enter a compare/update loop.

Ah, I see.  IIUC this won't happen because we're doing the DeepEqual not on a WorkloadEndpoint struct, but on the more limited WorkloadEndpointData struct.

The intention is that ever field in the `WorkloadEndpointData` will be both compared and also updated on the `WorkloadEndpoint`, so we won't hit that loop.

**Comment by bcreane** [comment]
ah, got it.

**Comment by bcreane** [comment]
then LGTM!

**Pull request merged**